core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,561 @@
+     * and ctx.entryState keeps the state.
+     *
+     * IMPORTANT:
+     *  1. Throws IllegalStateException if too much collisions are found
+     *
+     * @param ctx the context that will follow the operation following this key allocation
+     * @param key the key to write
+     * @param hashIdx hashIdx=fullHash||bit_size_mask
+     * @param fullHashIdx fullHashIdx=hashFunction(key)
+     * @return true only if the index is found (ctx.entryState keeps more details).
+     *         Otherwise (false), re-balance is required
+     *         ctx.entryIndex keeps the index of the chosen entry & ctx.entryState keeps the state
+     */
+    private boolean findSuitableEntryForInsert(ThreadContext ctx, K key, int hashIdx, long fullHashIdx) {
+        // start from given hash index
+        // and check the next `collisionChainLength` indexes if previous index is occupied
+        int collisionChainLengthLocal = collisionChainLength.get();
+        boolean entryFound = false;
+
+        // as far as we didn't check more than `collisionChainLength` indexes
+        for (int i = 0; i < collisionChainLengthLocal; i++) {
+            ctx.invalidate();
+            ctx.entryIndex = (hashIdx + i) % entriesCapacity; // check the entry candidate, cyclic increase
+            // entry's key is read into ctx.tempKey as a side effect
+            ctx.entryState = getEntryState(ctx, ctx.entryIndex, key, fullHashIdx);
+
+            // EntryState.VALID --> entry is occupied, continue to next possible location
+            // EntryState.DELETED_NOT_FINALIZED --> finish the deletion, then try to insert here
+            // EntryState.UNKNOWN, EntryState.DELETED --> entry is vacant, try to insert the key here
+            // EntryState.INSERT_NOT_FINALIZED --> you can compete to associate value with the same key
+            switch (ctx.entryState) {
+                case VALID:
+                    if (equalKeyAndEntryKey(ctx.key, key, hashIdx, fullHashIdx)) {
+                        // found valid entry has our key, the inserted key must be unique
+                        // the entry state will indicate that insert didn't happen
+                        return true;
+                    }
+                    continue;
+                case DELETED_NOT_FINALIZED:
+                    deleteValueFinish(ctx); //TODO: invocation must be within chunk publishing scope!
+                    // deleteValueFinish() also changes the entry state to DELETED
+                    // deliberate break through
+                case DELETED: // deliberate break through
+                case UNKNOWN:
+                    // deliberate break through
+                case INSERT_NOT_FINALIZED: // continue allocating value without allocating a key
+                    entryFound = true;
+                    break;
+                default:
+                    // For debugging the case new state is added and this code is not updated
+                    assert false;
```

Comment:
This will be a dead code.
To avoid it, you can add `assert ctx.entryState != null` before the `switch` clause and then remove the `default:` case.
This should compile well without errors or warnings.

Reply:
I do not see how this change protects us against new entry state being added without updating the switch... Let's discuss

Reply:
We didn't find how to protect from this feature possible bug. Leaving it as is meanwhile. Can be changed also later.